### ISSUE TITLE
Distance between beta images

### ISSUE THREAD

**Issue by fhopp** [cosanlab/nltools#393]
Hi there, 

more of a question than an issue with code, but perhaps of interest for the wider community. 
Say I want to compute the pairwise euclidean distance between subjects' beta maps (derived from a first-level GLM as described [here](https://dartbrains.org/content/Group_Analysis.html#running-a-group-analysis)). 
Per the API, the .distance method can "Calculate distance between images within a Brain_Data() instance."

So, is it as simple as loading all beta maps (contrasts in my case) into a Brain_Data class and then using the .distance method?

```
# Get all beta maps for a particular contrast
file_list = glob.glob(os.path.join(betas_dir, f'sub*_{con}*nii.gz'))
file_list.sort()
# Load files in a Brain_Data object
brain_dat = Brain_Data(file_list)
# Compute Euclidean distance between images and then convert distance matrix to similarity matrix
adj = brain_dat.distance(metric='euclidean').distance_to_similarity(metric="euclidean") 
``` 

This produced some rather odd-looking adjacency maps across the 64 subjects (perhaps an issue with individual subject maps  rather than the .distance method), but I wanted to double check that I am understanding the .distance method correctly. 

Thanks much! 
Freddy

 
![adj_mats](https://user-images.githubusercontent.com/22058606/120669310-2f944b00-c444-11eb-843a-733d2f83a7f4.png)

**Comment by ljchang** [comment]
Hi @fhopp, yes that's exactly right. The distance function will compute the pairwise distance (`using sklearn.metrics.pairwise_distance` function).  If you don't apply a mask it will be performing distance calculation on whole brain beta maps.  We tend to not use euclidean distance as a starting point with fMRI data unless we have thought through normalizing the data as often the intensities can vary quite a bit across participants.  Our default starting is correlation distance.  It definitely does look like you might have an outlier subject or two in some of your conditions. This definitely happens from people falling asleep, excessive head motion, or just being very different from anyone else. It's actually an issue we've been thinking through for awhile about how it can impact inferences using intersubject-representational similarity analysis.

**Comment by fhopp** [comment]
Hey @ljchang , 
thanks for the prompt response. IS-RSA is precisely what I am doing, although here for a moral judgment paradigm that is not time-locked and hence the GLM rather than classic ISC. Thanks for the simple pointer to the mask -- I will compute ROI based distances but also wanted a whole-brain map. 

It's a good point that correlational distance is less prone to variability in signal intensities. I quickly re-ran the above for correlational distance:
```
brain_sims = {}
for con in contrasts:
    # Get files for a particular contrast (e.g., carep-socn)
    file_list = glob.glob(os.path.join(betas_dir, f'sub*_{con}*nii.gz'))
    file_list.sort()
    # Store files in a Brain_Data object
    brain_dat = Brain_Data(file_list)
    # Compute correlational distance between each pair of subjects and then convert distance matrix to similarity matrix
    adj = brain_dat.distance(metric='correlation').distance_to_similarity(metric="correlation") 
    # Store matrix in dictionary
    brain_sims[con] = adj
```
This does produce less noisy maps (see output). I was mainly thinking about the [Walther](https://www.sciencedirect.com/science/article/pii/S1053811915011258?casa_token=8rf5cwDWljEAAAAA:RbLlNupyNdY1axosIH0EpwbRa-eUlxHAlDUBy_aVLRXtPJcsehtjb-fQgc0bWRwJzzpfzRmyeLA) paper and thus wanted to use euclidean over correlation distance. For comparison, I would like to run the same analysis again, but for euclidean distance with normalized BOLD as you suggest. Is there a principled normalization tool/pipeline that you recommend for this purpose? 

Thank you again, this was very helpful! 
![corr_adj_map](https://user-images.githubusercontent.com/22058606/120680943-bd296800-c44f-11eb-9b38-91bf94af2b56.png)

**Comment by ljchang** [comment]
i don't remember the specific details of the Walther paper, but in my experience with all of these things is that it's really hard to make sweeping rules that apply to all contexts.  In particular, RSA and IS-RSA although mathematically identical, differ in the distributions of values and many of the stats I do not believe generalize as well, though we haven't "proven" that yet.

We did ISRSA with GLM not ISC in this [decision-making paper](https://www.nature.com/articles/s41467-019-09161-6 ). Might be relevant for your moral judgement stuff.

**Comment by ljchang** [comment]
looks cool though, keep us posted!

**Comment by fhopp** [comment]
Will do! And yes, the decision-making paper is super cool and a good template for what I am trying to do. 
One last clarifying question regarding the distance calculation, which might be related to @ejolly 's #384 : 
When I use a correlation distance (1-r as expressed in [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html)), the diagonal should be 0s as there is a perfect correlation:
```
Adjacency(pairwise_distances(brain_dat.data, metric="correlation"), matrix_type="similarity")
``` 
Plotting the matrix shows that this is the case. 

However, when I do 1 - correlation distance, the diagonal should be 1:
```
Adjacency(1-pairwise_distances(brain_dat.data, metric="correlation"), matrix_type="similarity")
``` 
However, the produced diagonal are 0s (see also in your [RSA](https://dartbrains.org/content/RSA.html#single-subject-pattern-similarity) and [IS-RSA](https://naturalistic-data.org/content/Intersubject_RSA.html#calculate-brain-similarity) tutorials.) Do you know what is going on here/causing this behavior? Is the diagonal just not considered and I shouldnt worry about this? 

![image](https://user-images.githubusercontent.com/22058606/120725200-48275400-c48a-11eb-874c-ccef461dd4af.png)
